### DISCUSSION TITLE
How to add a column to model_has_roles and log with the sync() method?

### DISCUSSION THREAD

**Discussion by alexv96** [spatie/laravel-permission#1906] [Q&A] [answered]
I am working with a SaaS system where I have the users, roles and permissions in the parent database (Landlord) and I would like to create the tenant_id column because it does not allow me to integrate only the model_has_roles table in the tenant databases (Tenant), then a solution that occurred to me was to create in the table the tenant column, but I do not know how to keep the sync methods and that I register in that column.

**Comment by erikn69** [comment, marked as answer]
https://github.com/spatie/laravel-permission/discussions/1827#discussioncomment-1234029 add field to migration
overwrite relations or use [teams-permissions](https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions)

**Reply by erikn69** [reply]
@alexv96 you have to add a middleware to put always `team_id` with `setPermissionsTeamId($model);` before start working with permissions package. read [documentation](https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions#working-with-teams-permissions)

**Reply by alexv96** [reply]
I did the issue that now I want to access the roles by user and in the traditional way was Auth::user()->roles but currently does not bring me the roles to make the query, I manage to make it register in the database but nothing happens does not bring me the roles that I assign to users.

Currently I use this form in a table because before using teams it worked for me, ```{{ implode( ', ' , $user->roles()->pluck('name')->toArray() ) }} }}```

And in the documentation it doesn't say anything about changing the way to call when you have teams.

**Reply by erikn69** [reply]
>And in the documentation it doesn't say anything about changing the way to call when you have teams.

Read again

**Reply by alexv96** [reply]
I read again and it only talks about assign and remove, but I want to list all users with their relevant roles,

**Reply by erikn69** [reply]
>only talks about assign and remove

it's the same logic 

You have to work one team at the time per session, and do assign and read data only for that session
Look at the test, it's working
https://github.com/spatie/laravel-permission/blob/1c9304af746b42308a1d1da2989e12a1439571a9/tests/TeamHasRolesTest.php#L74-L80
![image](https://user-images.githubusercontent.com/4933954/140094016-4ec2cb0c-4bba-4472-a20a-76a7b9938717.png)
![image](https://user-images.githubusercontent.com/4933954/140094123-b1892cd9-5797-4ca7-b806-7cf46c6336b5.png)

**Reply by alexv96** [reply]
Ok, reading again and from what you say, I have the middleware configured.

The model I can call it as it is and not extend it with any kind of spatie, then from that model I think I should make the call?

Because for the moment I followed the example and I do not achieve it :'( I tried to create it all with middleware and as it should be explained but it does not happen I continue in the same thing, perhaps I should call it from the new class that is created in the tutorial?

Look this is what I did.

```
**New Middleware**
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Cookie;

class TeamsPermission
{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle(Request $request, Closure $next)
    {
        if (!empty(Auth::user())) {
            app(\Spatie\Permission\PermissionRegistrar::class)->setPermissionsTeamId(\Cookie::get('tenant'));
        }
        return $next($request);
    }
}
```

I tried to force it through the team but I didn't get anything.

```
@php
    $a = getPermissionsTeamId();
    $b = Auth::user()->setPermissionsTeamId(7);
    dd($a,$b);
@endphp
```

![image](https://user-images.githubusercontent.com/33232796/140163902-d8491a40-462a-47cb-adde-b89341f4ca7e.png)

**Reply by erikn69** [reply]
Questions,
did you add that middleware to kernel?
Why `Auth::user()->setPermissionsTeamId(7)`? must be only `setPermissionsTeamId(7)` and don´t returns anything
Put `\Cookie::get('tenant')` on `dd()` too
Also put your middleware at last, after cookie reading
I am using teams like others without problems

**Reply by alexv96** [reply]
**Q: do you added that middleware to kernel?**
A: I did add it, I ran a dd() to see if it is working, but it doesn't enter the if when I log in.

I also added it to the kernel but it only enters if I remove the exclamation mark or put an else.

**Q: Why Auth::user()->setPermissionsTeamId(7)? must be only setPermissionsTeamId(7) and don´t returns anything**
A: Because I was trying to somehow force the display of user roles. 😅 Here I forced with 7 because I was testing in another project that I created a blank one to see this issue to see if there was any other library that made it not work.

**Q: Put \Cookie::get('tenant') on dd() too**
A: The dd() is to test what data it is bringing me that I am testing it directly from blade.

Cookie::get('tenant') returns the team number.

![image](https://user-images.githubusercontent.com/33232796/140170377-a3dfa833-c0d2-4922-a104-fdfb35ea385e.png)

The difference is that the middlewarePriority in my kernel file was never in Laravel 8 so I read that I had to put it and I put it, maybe that generates that I can't access?

**This is my kernel TeamsPermission**
```
<?php

namespace App\Http;

use Illuminate\Foundation\Http\Kernel as HttpKernel;

class Kernel extends HttpKernel
{
    /**
     * The application's global HTTP middleware stack.
     *
     * These middleware are run during every request to your application.
     *
     * @var array
     */
    protected $middleware = [
        // \App\Http\Middleware\TrustHosts::class,
        \App\Http\Middleware\TrustProxies::class,
        \Fruitcake\Cors\HandleCors::class,
        \App\Http\Middleware\PreventRequestsDuringMaintenance::class,
        \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
        \App\Http\Middleware\TrimStrings::class,
        \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
        // \App\Http\Middleware\TeamsPermission::class,

    ];

    /**
     * The application's route middleware groups.
     *
     * @var array
     */
    protected $middlewareGroups = [
        'web' => [
            \App\Http\Middleware\EncryptCookies::class,
            \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
            \Illuminate\Session\Middleware\StartSession::class,
            // \Illuminate\Session\Middleware\AuthenticateSession::class,
            \Illuminate\View\Middleware\ShareErrorsFromSession::class,
            \App\Http\Middleware\VerifyCsrfToken::class,
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],

        'api' => [
            // \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
            'throttle:api',
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],

        'tenant' => [
            \Spatie\Multitenancy\Http\Middleware\NeedsTenant::class,
            \Spatie\Multitenancy\Http\Middleware\EnsureValidTenantSession::class,
        ],
    ];

    protected $middlewarePriority = [
        \App\Http\Middleware\TeamsPermission::class,

    ];
}
```

**Reply by erikn69** [reply]
Middleware priority it's for set order of the middleware, if you put only one it does nothing, you must set it to web or your middleware group, and do you have another guards? I'm on a cell phone and I can't read everything, but looks like you don't know how to config your kernel

**Reply by alexv96** [reply]
Don't worry when you can.

I'm sorry that I don't know much about middleware and kernel in Laravel, it's something very basic in this part.

Now I can enter the middleware IF without problems, but I still can not list the roles of all users, or in this case to know what roles I have assigned as a user.

Because in the database the team is registered in the relevant tables.

**Reply by erikn69** [reply]
Here is my config, but i am using session instead of cookie, when you do use `<>`  for code, set it to php for highlighting and readability
```php
protected $middleware = [
    \App\Http\Middleware\TrustProxies::class,
    \App\Http\Middleware\CheckForMaintenanceMode::class,
    \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
    \App\Http\Middleware\TrimStrings::class,
    \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
    \Fruitcake\Cors\HandleCors::class,
];
protected $middlewareGroups = [
    'web' => [
        \App\Http\Middleware\EncryptCookies::class,
        \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
        \Illuminate\Session\Middleware\StartSession::class,
        \Illuminate\View\Middleware\ShareErrorsFromSession::class,
        \App\Http\Middleware\VerifyCsrfToken::class,
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
        \App\Http\Middleware\TeamsPermission::class,
    ],

    'api' => [
        'throttle:60,1',
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
        \App\Http\Middleware\TeamsPermission::class,
    ],
];
protected $middlewarePriority = [
    \App\Http\Middleware\EncryptCookies::class,
    \Illuminate\Session\Middleware\StartSession::class,
    \Illuminate\View\Middleware\ShareErrorsFromSession::class,
    \App\Http\Middleware\Authenticate::class,
    \Illuminate\Session\Middleware\AuthenticateSession::class,
    \Illuminate\Routing\Middleware\SubstituteBindings::class,

    \App\Http\Middleware\TeamsPermission::class,
    \Illuminate\Auth\Middleware\Authorize::class,

    \Spatie\Permission\Middlewares\RoleMiddleware::class,
    \Spatie\Permission\Middlewares\PermissionMiddleware::class,
    \Spatie\Permission\Middlewares\RoleOrPermissionMiddleware::class,
];
```

**Reply by alexv96** [reply]
AAAA that's why it didn't work, now everything works, thank you very much, this middleware was new, it didn't appear in any tutorial I had seen, thank you very much. 

Thank you very much, now everything works.
